diff --git a/Gemfile b/Gemfile
index 153dd3d..96d0454 100644
--- a/Gemfile
+++ b/Gemfile
@@ -8,12 +8,16 @@ group :development do
   gem 'jekyll-sitemap'
   gem 'rdiscount', '~> 2.0'
   gem 'RedCloth', '~> 4.2.9'
+  gem 'coderay', '~> 1.1.0'
   gem 'haml', '~> 4.0'
   gem 'compass', '~> 1.0.1'
   gem 'sass-globbing', '~> 1.0.0'
   gem 'rubypants', '~> 0.2.0'
   gem 'rb-fsevent', '~> 0.9'
   gem 'stringex', '~> 1.4.0'
+  gem 'thin', '~> 1.6.3'
+  gem 'gemoji', '~> 2.1.0'
+  gem 'json', '~> 1.8.1'
 end
 
 gem 'sinatra', '~> 1.4.2'
