diff --git a/master:Gemfile b/source:Gemfile
index 9557d59..c28183a 100644
--- a/master:Gemfile
+++ b/source:Gemfile
@@ -8,12 +8,15 @@ group :development do
   gem 'jekyll-sitemap'
   gem 'rdiscount', '~> 2.0'
   gem 'RedCloth', '~> 4.2.9'
+  gem 'coderay'
   gem 'haml', '~> 4.0'
   gem 'compass', '~> 0.12.2'
   gem 'sass-globbing', '~> 1.0.0'
   gem 'rubypants', '~> 0.2.0'
   gem 'rb-fsevent', '~> 0.9'
   gem 'stringex', '~> 1.4.0'
+  gem 'thin'
+  gem 'gemoji'
 end
 
 gem 'sinatra', '~> 1.4.2'
