diff --git a/Gemfile b/Gemfile
index 5683812..599353f 100644
--- a/Gemfile
+++ b/Gemfile
@@ -8,11 +8,23 @@ group :development do
   gem 'jekyll-sitemap'
   gem 'rdiscount', '~> 2.0'
   gem 'RedCloth', '~> 4.2.9'
+  gem 'coderay', '~> 1.1.0'
   gem 'haml', '~> 4.0'
   gem 'compass', '~> 1.0.1'
   gem 'sass-globbing', '~> 1.0.0'
   gem 'rb-fsevent', '~> 0.9'
   gem 'stringex', '~> 1.4.0'
+  gem 'thin'
+  gem 'gemoji'
+  gem 'json'
+  gem 'octopress-asset-pipeline'
+  gem 'yui-compressor'
+  gem 'htmlcompressor'
+  gem 'parallel'
+  gem 'chronic'
+  gem 'google-api-client'
+  gem 'ruby-prof'
+  gem 'ruby-progressbar'
 end
 
 gem 'sinatra', '~> 1.4.2'
