diff --git a/plugins/category_generator.rb b/plugins/category_generator.rb
index 769a55b..18c1ba2 100644
--- a/plugins/category_generator.rb
+++ b/plugins/category_generator.rb
@@ -153,7 +153,16 @@ ERR
     # Returns string
     #
     def category_links(categories)
-      categories.sort.map { |c| category_link c }.join(', ')
+      categories = categories.sort!.map { |c| category_link c }
+
+      case categories.length
+      when 0
+        ""
+      when 1
+        categories[0].to_s
+      else
+        "#{categories[0...-1].join(', ')}, #{categories[-1]}"
+      end
     end
 
     # Outputs a single category as an <a> link.
@@ -164,7 +173,7 @@ ERR
     #
     def category_link(category)
       dir = @context.registers[:site].config['category_dir']
-      "<a class='category' href='/#{dir}/#{category.to_url}/'>#{category}</a>"
+      "<a class='category' itemprop='articleSection keywords' href='/#{dir}/#{category.to_url}/'>#{category}</a>"
     end
 
     # Outputs the post.date as formatted html, with hooks for CSS styling.
@@ -174,8 +183,8 @@ ERR
     # Returns string
     def date_to_html_string(date)
       result = '<span class="month">' + date.strftime('%b').upcase + '</span> '
-      result << date.strftime('<span class="day">%d</span> ')
-      result << date.strftime('<span class="year">%Y</span> ')
+      result += date.strftime('<span class="day">%d</span> ')
+      result += date.strftime('<span class="year">%Y</span> ')
       result
     end
 
diff --git a/plugins/image_tag.rb b/plugins/image_tag.rb
index 4567000..15adef5 100644
--- a/plugins/image_tag.rb
+++ b/plugins/image_tag.rb
@@ -24,7 +24,7 @@ module Jekyll
     def initialize(tag_name, markup, tokens)
       attributes = ['class', 'src', 'width', 'height', 'title']
 
-      if markup =~ /(?<class>\S.*\s+)?(?<src>(?:https?:\/\/|\/|\S+\/)\S+)(?:\s+(?<width>\d+))?(?:\s+(?<height>\d+))?(?<title>\s+.+)?/i
+      if markup =~ /(?<class>\S.*\s+)?(?<src>(?:https?:\/\/|\S*\/|val:)\S+)(?:\s+(?<width>\d+))?(?:\s+(?<height>\d+))?(?<title>\s+.+)?/i
         @img = attributes.reduce({}) { |img, attr| img[attr] = $~[attr].strip if $~[attr]; img }
         if /(?:"|')(?<title>[^"']+)?(?:"|')\s+(?:"|')(?<alt>[^"']+)?(?:"|')/ =~ @img['title']
           @img['title']  = title
@@ -39,6 +39,12 @@ module Jekyll
 
     def render(context)
       if @img
+        @img['src'] = context[@img['src'].split(':')[1]] if @img['src'] =~ /^val:/
+        if ! (@img['class'] and @img['class'].include?('noimgpath'))
+          if context['site.imgpath'] and @img['src'] !~ /^(http|#{context['site.imgpath']})/
+            @img['src'] = context['site.imgpath']+@img['src']
+          end
+        end
         "<img #{@img.collect {|k,v| "#{k}=\"#{v}\"" if v}.join(" ")}>"
       else
         "Error processing input, expected syntax: {% img [class name(s)] [http[s]:/]/path/to/image [width [height]] [title text | \"title text\" [\"alt text\"]] %}"
diff --git a/plugins/include_array.rb b/plugins/include_array.rb
index 000040f..5c744c2 100644
--- a/plugins/include_array.rb
+++ b/plugins/include_array.rb
@@ -31,6 +31,11 @@ module Jekyll
       end
 
       rtn = ''
+      n_files = context.environments.first['site'][@array_name].size ? \
+        context.environments.first['site'][@array_name].size : 0
+      n_files_count = n_files
+      n_last_fixed = context.registers[:site].config['n_scroll_fixed'] ? \
+        context.registers[:site].config['n_scroll_fixed'] : 0
       (context.environments.first['site'][@array_name] || []).each do |file|
         if file !~ /^[a-zA-Z0-9_\/\.-]+$/ || file =~ /\.\// || file =~ /\/\./
           rtn = rtn + "Include file '#{file}' contains invalid characters or sequences"
@@ -42,11 +47,19 @@ module Jekyll
             source = File.read(file)
             partial = Liquid::Template.parse(source)
             context.stack do
+              if n_files_count == n_last_fixed
+                rtn = rtn + "<div class=\"scroll-fixed\">"
+              end
+
               rtn = rtn + partial.render(context)
+              if n_last_fixed != 0 and n_files_count == 1
+                rtn = rtn + "</div>"
+              end
             end
           else
             rtn = rtn + "Included file '#{file}' not found in _includes directory"
           end
+          n_files_count -= 1
         end
       end
       rtn
diff --git a/plugins/octopress_filters.rb b/plugins/octopress_filters.rb
index 48413a8..5e75176 100644
--- a/plugins/octopress_filters.rb
+++ b/plugins/octopress_filters.rb
@@ -56,11 +56,28 @@ module OctopressLiquidFilters
     end
   end
 
+  def excerpt_after(input)
+    if input.index(/<!--\s*more\s*-->/i)
+      input.split(/<!--\s*more\s*-->/i)[1]
+    else
+      ""
+    end
+  end
+
   # Checks for excerpts (helpful for template conditionals)
   def has_excerpt(input)
     input =~ /<!--\s*more\s*-->/i ? true : false
   end
 
+  # Get first amazon-img
+  def amazon_img(input)
+    if input.match(/<div class="amazon-img">.*<\/div>/) != nil
+      input.match(/<div class="amazon-img">.*<\/div>/)[0].sub("_SS200_", "_SS90_CR0,0,120,90")
+    else
+      #"no amazon-img"
+    end
+  end
+
   # Summary is used on the Archive pages to return the first block of content from a post.
   def summary(input)
     if input.index(/\n\n/)
@@ -73,6 +90,11 @@ module OctopressLiquidFilters
   # Extracts raw content DIV from template, used for page description as {{ content }}
   # contains complete sub-template code on main page level
   def raw_content(input)
+    /<div class="entry-content">(?<content>[\s\S]*?)<!--\s*more\s*-->/ =~ input
+    if ! (content.nil?)
+      return content
+    end
+
     /<div class="entry-content">(?<content>[\s\S]*?)<\/div>\s*<(footer|\/article)>/ =~ input
     return (content.nil?) ? input : content
   end
@@ -85,7 +107,7 @@ module OctopressLiquidFilters
   # Replaces relative urls with full urls
   def expand_urls(input, url='')
     url ||= '/'
-    input.gsub /(\s+(href|src|poster)\s*=\s*["|']{1})(\/[^\/>]{1}[^\"'>]*)/ do
+    input.gsub /(\s+(href|src)\s*=\s*["|']{1})(\/[^\/>]{1}[^\"'>]*)/ do
       $1+url+$3
     end
   end
